### ISSUE TITLE
Very slow read speed on the qemu image

### ISSUE THREAD

**Issue by Eboubaker** [cyberkernelofficial/docker-in-termux#13]
I noticed massive read speed(like 10-20 times) reduction when inside qemu this makes building/pulling images take hours.
Do you have any improvements that can be made here?

I tested with this script https://github.com/thodnev/MonkeyTest/blob/master/monkeytest.py

**Comment by cyberkernelofficial** [comment]
You're right, running Docker within QEMU can significantly impact disk read/write speeds, leading to longer build and pull times. This is often due to the virtualization overhead and how QEMU emulates disk access.

While using a `.raw` disk image instead of QCOW2 can yield some improvement, the core issue often lies in the limited resources available to QEMU. As suggested in [this issue](https://github.com/cyberkernelofficial/docker-in-termux/issues/11), allocating more CPU cores and memory to your QEMU virtual machine can significantly boost performance. 

> Additionally, ensure that your host system (where you're running QEMU) has sufficient resources available. If the host system is already under heavy load, it will further impact QEMU's performance.

Here are some things you can try:
>* Increase CPU Cores: Allocate more CPU cores to the QEMU virtual machine. Experiment to find the optimal number for your system.
>* Increase Memory: Provide QEMU with more RAM. Building Docker images can be memory-intensive.
>* Experiment with QEMU Settings:  Explore QEMU's configuration options related to disk I/O and caching to see if any optimizations can be made. 

Let me know if these suggestions help improve your build speeds within QEMU. If you find any specific configurations that work well, please share them with the community so others can benefit!

**Comment by Eboubaker** [comment]
I feel sad that you had to use AI to write this. However I hope someone can provide configuration that can mitigate the speed issue because this trick is completely useless to run mid-large docker containers.

**Comment by cyberkernelofficial** [comment]
I understand your concern about AI use in this project.  While I do leverage AI to improve my writing (as a non-native English speaker) and make my communication clearer, I assure you that the core functionality of this repo is not solely reliant on AI-generated code. 

I appreciate your feedback on the speed issue. It's definitely a valid concern, and I'm actively working on optimizing the configuration to improve performance for larger Docker containers. I'm open to suggestions and contributions from the community. 

Let's work together to make this project as efficient and helpful as possible!

Note: this is AI generated too

**Comment by freebrowser1** [comment]
Maybe use arm64 architecture instead of x86-64. The latter is emulated on Android and ARM is native speed.

**Comment by Eboubaker** [comment]
> Maybe use arm64 architecture instead of x86-64. The latter is emulated on Android and ARM is native speed.

How can you do that in the terminal exactly I couldn't make it run.
this is what I tried
```
pkg install qemu-system-arm
qemu-img create -f raw alpine-arch64.img 2G
wget https://dl-cdn.alpinelinux.org/alpine/v3.20/releases/aarch64/alpine-virt-3.20.0-aarch64.iso
qemu-system-arm -machine virt -m 1024 -smp cpus=2 -drive if=pflash,format=raw,read-only=on,file=$PREFIX/share/qemu/edk2-aarch64-code.fd -netdev user,id=n1,dns=8.8.8.8,hostfwd=tcp::2222-:22 -device virtio-net,netdev=n1 -drive file=alpine-arch64.img,format=raw,if=virtio -cdrom alpine-virt-3.20.0-aarch64.iso -nographic
```
But it just does nothing I waited 30 mins

**Comment by freebrowser1** [comment]
Typo ?
alpine-arch64 should be alpine-**aa**rch64.

**Comment by Eboubaker** [comment]
> Typo ? alpine-arch64 should be alpine-**aa**rch64.

There is no typo it's the disk filename that I have created with `qemu-img create -f raw alpine-arch64.img 2G`

**Comment by freebrowser1** [comment]
You can try the link above from hub.docker.com and pick the second entry which is aarch64.

**Comment by freebrowser1** [comment]
Is this not a Docker image for Termux which you can deploy under Termux on a non-rooted Android ?

**Comment by Eboubaker** [comment]
> Is this not a Docker image for Termux which you can deploy under Termux on a non-rooted Android ?

No, Try to read or translate the README.md file https://github.com/cyberkernelofficial/docker-in-termux/blob/main/README.md
The purpose is to install Qemu emulator so that we can run any docker image inside it. (you can't install docker in non-rooted termux)

**Comment by Eboubaker** [comment]
Ok I have reached to the Mailing list of Qemu to get help in regards to running it in aarch64, I have made a mistake by using `qemu-system-arm` which is aarch32bit I had to use `qemu-system-aarch64` it can run arm64 bit.
```shell
pkg install qemu-system-aarch64
qemu-img create -f raw alpine-arch64.img 2G
wget https://dl-cdn.alpinelinux.org/alpine/v3.20/releases/aarch64/alpine-virt-3.20.0-aarch64.iso
qemu-system-aarch64 -cpu cortex-a55 \
-machine virt \
-m 1024 \
-smp cpus=2 \
-drive if=pflash,format=raw,read-only=on,file=$PREFIX/share/qemu/edk2-aarch64-code.fd \
-netdev user,id=n1,dns=8.8.8.8,hostfwd=tcp::2222-:22 \
-device virtio-net,netdev=n1 \
-cdrom alpine-virt-3.20.0-aarch64.iso -nographic alpine-arch64.img
```
> I used cpu cortex-a55 here because I checked supported architectures of my phone using CPU-Z app, then I checked the available list using `qemu-system-aarch64 -cpu help` because I wanted them to match, I don't know if that helps with performance though...

When I did the benchmark again nothing changed... I still get 4Mb/s read speed on the disk which is disappointing. Thanks anyways @freebrowser1 for the idea.

**Comment by freebrowser1** [comment]
The problem is that the arm64 variant is also emulated (which is slow) rather than virtualized. Most Android devices require hardware virtualization which may require root.

**Comment by Eboubaker** [comment]
> The problem is that the arm64 variant is also emulated (which is slow) rather than virtualized. Most Android devices require hardware virtualization which may require root.

The problem is in I/O performance, If we can figure a way to boost disk speed half of the problem is fixed.

**Comment by Eboubaker** [comment]
I will just leave this here

https://www.androidauthority.com/android-linux-terminal-app-3489887/


https://www.androidauthority.com/android-linux-terminal-purpose-3535765/

**Comment by bluefing** [comment]
alpine:~# python monkey.py
Reading: 100.00 %

Written 128 MB in 0.6622 s
Write speed is  193.30 MB/s
  max: 219.29, min: 78.79

Read 262144 x 512 B blocks in 6.8591 s
Read speed is  18.66 MB/s
  max: 25.34, min: 0.21

Brought to you by coding monkeys.
Eat bananas, drink coffee & enjoy!


FWIW, performance on pixel 10 pro

**Comment by susu3132** [comment]
Hi, have you tested the Raw + Loop mode? Disguising the Raw file as a real physical disk (block device) can result in shorter I/O path.

**Comment by susu3132** [comment]
Haha, still not working (ಥ_ಥ). p.s, this read speed is ridiculously slow.
moto razr 2022

termux：
.../files/home # python3 monkeytest.py
Reading: 100.00 %

Written 128 MB in 0.3012 s
Write speed is  424.97 MB/s
  max: 598.48, min: 85.40

Read 262144 x 512 B blocks in 0.6127 s
Read speed is  208.92 MB/s
  max: 323.15, min: 3.06

qemu：
alpine:~# python3 monkeytest.py
Reading: 100.00 %

Written 128 MB in 1.7838 s
Write speed is  71.76 MB/s
  max: 101.83, min: 12.68

Read 262144 x 512 B blocks in 43.6478 s
Read speed is  2.93 MB/s
  max: 7.30, min: 0.08
